articles/data-classes-in-python/index.md

```diff
@@ -0,0 +1,211 @@
+
+### Using default values
+Data classes also allow us to define the default value of our attributes when they are declared. The snippet below shows how to go about this.
+
+```python
+from dataclasses import dataclass, field
+import random
+
+def gpa_range():
+    return float(random.randrange(2, 5))
+
+
+@dataclass
+class Student:
+    name: str = "No Name"
+    course: str = "No Course"
+    gpa: float = field(default_factory=gpa_range)
+
+# create a default student object
+student1 = Student()
+print(student1)
+
+# Create a specified book, gpa is set by field operator
+student1 = Student('James', 'Comp Science')
+student2 = Student('Angie', 'Bcom')
+
+print(student1)
+print(student2)
+```
+
+In our class object, we indicate the default value by using an equal sign with the corresponding data type. We pass `No Name` and `No Course` for the string data types and pass a field with a defined function `gpa_range()` to generate random GPA values to the gpa attribute. 
+
+The `default_factory` allow us to use mutable default values using the `field()` specifier. The `gpa_range()` function generate random values between 2 and 5. We then create an instance of Student class by passing the `name` and `course` attribute. The `gpa` property is set by the field operator which we imported. Running the file outputs a Student object with all the properties.
+
+### Immutable data classes
+Immutable data classes are classes that do not allow us to change their values and throws an exception when modified. Why use immutable classes? 
+
+Well, first, immutable classes are thread-safe since their content cannot be changed. It also improves clarity and accuracy when reading one's code. The snippet below explains how to go about it.
+
+```python
+from dataclasses import dataclass
+
+@dataclass(frozen=True)
+class ImmutableDataClass:
+    val1: str = "This value"
+    val2: int = 0
+
+obj = ImmutableDataClass()
+print(obj.val1)
+
+obj.val1 = "Another value"
+print(obj.val1)
+```
+
+We create a new class, `ImmutableDataClass()` and define two attributes: `val1` and `val2`. We assign default values to both: `This value` and `0` respectively. We create a variable `obj` and print `val1`. `"This value"` is displayed. If we try to modify `val1` by assigning it another value, `" Another value"` we get an exception, `FrozenInstanceError`. This is because the class is immutable. In our decorator we pass `Frozen=True` to the `dataclass()` decorator. This make the python objects to be immutable. 
+
+The frozen classes can't modify themselves either. For instance the `func()` function in the `ImmutableDataClass` cannot be modified too. 
+
```

Comment:
@OkothPius Running the code below throws an exception at the line: `val1: str = "This value"`. Please modify your code to throw an exception at `func()`